accepted/0000-pure-mode.md

```diff
@@ -0,0 +1,344 @@
+# Pure mode
+
+## Summary
+
+This RFC is a proposal to add a new opt-in installation mode.
+
+This mode would be the missing ingredient to make workspaces truly scalable.
+
+## Motivation
+
+The introduction of workspaces in npm v7 brought the ability to split large codebases into smaller pieces of code with declared dependencies between them.
+
+Several build tools took advantage of these declared dependencies to split the builds into multiple steps (one per workspace) which could be cached or run in parallel. This caching and parallelism usually improves significantly build performance.
+
+The current default installation strategy does not properly communicates to Node.js the information about the dependency graph. Node.js having an inaccurate view of the dependency graph leads the build systems to wrongly skip build operations, leading to risk of bugs slipping through.
+
+
+The installation mode proposed by this RFC would communicate accurately to Node.js the dependency graph, allowing build tools to correctly reason about it.
+
+## Naming
+
+The proposed name for this mode is `pure-mode`, meaning "pure" as in "no side effects". This name comes from the parallel with functional programming; in functional programming, functions are called `pure` because they don't have side effects, in `pure-mode` workspaces have no side effects.
```

Comment:
node module resolution does and *should* mean that a parent package.json's node_modules should be requireable in a child - so i don't think "no side effects" is actually what this would be.

Reply:
The suggested `pure-mode` requires a package to declare a `peer-dependency` in order to be able to impact the dependencies of its parent. Importing code from one of your siblings without having a peer-dependency declared contributes to the problems that this RFC is trying to solve.

Reply:
I don’t think that’s sufficient - users will always want every child package to be shared also, whether it’s a peer or not.

Reply:
Can you clarify the scenario that this mode needs to support?

Is there any data (article, discussion, blog post...) about users expecting to have access to parent's dependencies?

Reply:
That’s how node has always worked - every use of `require` expects that. That’s not what we’re talking about tho.

I’m claiming that what users want is to have both peers and siblings automatically shared, when matching. Currently, lerna/yarn/npm does this via hoisting, which has tons of drawbacks. I think what we need is a way to do it *not* solely with hoisting.

No matter what, a parent package.json’s node_modules must always be requireable because that’s how node works.

Reply:
My assumption is that the fact that a package can require its siblings is a side effect of the package managers implementation, not a requirement described anywhere. Or is it specified somewhere?

pnpm and yarn PnP do work without fulfilling this expectation so I would guess that most packages are not relying on it. For the ones that do, it would be very easy to fix by declaring a peerDependency to properly express this expectation.

Reply:
Absent workspaces, a package can not require its siblings' deps, but it can always require its parent's deps.

PnP has to patch a number of parts of the ecosystem (in its pnp.js file) to ensure things don't break, and both PnP and pnpm have nonzero bugs caused by their approach. I think that it would be a very bad idea for npm to apply this approach, *especially* only on workspaces.

Reply:
Discussed during the RFC call and @ljharb said that this is not a concern for him anymore.